modules/accredited_representative_portal/app/controllers/accredited_representative_portal/v0/form21a_controller.rb

```diff
@@ -5,9 +5,18 @@ module V0
     # Form21aController handles the submission of Form 21a to the accreditation service.
     # It parses the request body, submits the form via AccreditationService, and processes the response.
     class Form21aController < ApplicationController
+      class SchemaValidationError < StandardError
+        attr_reader :errors
+
+        def initialize(errors)
+          @errors = errors
+          super("Validation failed: #{errors}")
+        end
+      end
+
       FORM_ID = '21a'
 
-      before_action :parse_request_body, only: [:submit]
+      before_action :parse_request_body, :validate_schema, only: [:submit]
```

Comment:
I spent a little time playing with the control flow and the fact that `before_action`'s can [halt the action as well as following `before_action`'s when they render (or redirect)](https://api.rubyonrails.org/v7.1.5/classes/AbstractController/Callbacks/ClassMethods.html#method-i-before_action) does actually express the control flow really nicely. But I think we should add some protection to someone accidentally changing the order of these `before_action`'s. Maybe a class method called `validate_form_before_submit` that includes a comment explaining what Rails is doing for us for control flow here + the importance of the order of the `before_action`'s? Another option could be an array constant with such a comment but that seems a little funkier?

Did you have this control flow question in mind to arrive at this?

Reply:
A couple more options (opposite ends of the spectrum) are 1) just putting a comment above this line, 2) mixing in a module that encapsulates these two methods + a helper class method.

Reply:
Thanks for the thoughtful review! Great points.  I think the order dependency here is actually self-enforcing - `validate_schema` would fail immediately if someone tried to run it before `parse_request_body` since there'd be no parsed data to validate. Adding the encapsulating class method might be overkill?

I personally would opt for the simplest option you laid out and add some explanatory comment at least for now. If more filters (or even controllers) are added and this becomes a bigger issue then I think your ideas make a lot of sense.

However I'd defer to you so whatever you think is best at this stage.

Reply:
Just a comment seems fine.